Add document:updateByQuery (#1559)
kuzzleio/kuzzle: doc/2/api/controllers/document/update-by-query/index.md

---
code: true
type: page
title: updateByQuery
---

# updateByQuery

Updates documents matching the provided search query. 

> Documents updated that way trigger real-time notifications.

Reviewer: IMHO it's not necessary to specify that this method will trigger realtime notifications because all API method should trigger notifications

Author: It's a reminder, and I wrote it to be consistent with other method like
https://docs.kuzzle.io/core/2/api/controllers/document/delete-by-query/

## Limitations

The request fails if the number of documents returned by the search query exceeds the `documentsWriteCount` server configuration (see the [Configuring Kuzzle](/core/2/guides/essentials/configuration) guide).

To update a greater number of documents, either change the server configuration, or split the search query.

---

## Query Syntax

### HTTP

```http
URL: http://kuzzle:7512/<index>/<collection>/_queryUpdate

> Method: POST

Reviewer 2: Shouldn't this route use the HTTP PUT verb as the `document:update` method?

Body:
```

```js
{
  "query": {
    // ...
  }
}
```

### Other protocols

```js
{
  "index": "<index>",
  "collection": "<collection>",
  "controller": "document",
  "action": "updateByQuery",
  "body": {
    "query": {
      // query to match documents
    },
    "changes": {
      // documents changes
    }
  }
}
```

---

## Arguments

- `collection`: collection name
- `index`: index name

---

### Optional:

- `source`: if set to `true` Kuzzle will return the updated documents body in the response.

## Body properties

> - `query`: documents matching this search query will be updated. Uses the [ElasticSearch Query DSL](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/query-dsl.html) syntax.

Reviewer: Missing the `changes` body property

---

> ## Response

Reviewer: I think it will be better to names the 2 arrays `successes` and `errors` like for the `m*` methods of the API.  

See https://docs.kuzzle.io/core/2/api/controllers/document/m-update/

Also, the `total` property is useless since you just have to compute the `successes` array length.

> Returns a `result` object containing the a `documents` array of the updated documents.

Reviewer: ```suggestion
Returns a `result` object containing a `documents` array representing updated documents.
```

```js
{
  "status": 200,
  "error": null,
  "index": "<index>",
  "collection": "<collection>",
  "controller": "document",
  "action": "updateByQuery",
  "requestId": "<unique request identifier>",
  "result": {
    "documents": [
      {

>         _id: "document-1",

Reviewer: ```suggestion
        "_id": "document-1",
```
All properties must be enclosed with double quote in valid JSON

>         _source: [Object],  // If `source` option is set to true

Reviewer: Nitpicking (same format as for the `document:update` route)
```suggestion
        _source: "<updated document>" // If `source` option is set to true
```

        status: 200,

>         result: "updated"

Reviewer: ```suggestion
        result: "updated"
```
Seems unnecessary to me since this method can only update document (not create them) and failures will end in the `failures` array

---
code: true
type: page
title: updateByQuery
---

# updateByQuery

Updates documents matching the provided search query. 

Documents updated that way trigger real-time notifications.

## Limitations

The request fails if the number of documents returned by the search query exceeds the `documentsWriteCount` server configuration (see the [Configuring Kuzzle](/core/2/guides/essentials/configuration) guide).

To update a greater number of documents, either change the server configuration, or split the search query.

---

## Query Syntax

### HTTP

```http

> URL: http://kuzzle:7512/<index>/<collection>/_queryUpdate

Reviewer: After thinking, just a `PUT` request on `/_query` would be better to keep consistency with deleteByQuery 
```suggestion
URL: http://kuzzle:7512/<index>/<collection>/_query
```

Method: PUT
Body:
```

```js
{
  "query": {
    // query to match documents
  },
  "changes": {
    // documents changes
  }
}
```

### Other protocols

```js
{
  "index": "<index>",
  "collection": "<collection>",
  "controller": "document",
  "action": "updateByQuery",
  "body": {
    "query": {
      // query to match documents
    },
    "changes": {
      // documents changes
    }
  }
}
```

---

## Arguments

- `collection`: collection name
- `index`: index name

---

> ### Optional:

Reviewer 3: Missing `refresh` option
